File: tests/test_visual_check.js
#!/usr/bin/env node
/**
 * Visual testing script for Harada Method application.
 * Tests landing page and chart view for visual elements.
 */

const puppeteer = require('puppeteer');

async function testLandingPage(page) {
  console.log('\n=== Testing Landing Page ===');
  const issues = [];
  
  // Navigate to landing page
  await page.goto('http://localhost:3000', { waitUntil: 'networkidle0' });
  
  // Take screenshot
  await page.screenshot({ path: 'landing_page.png', fullPage: true });
  console.log('✓ Screenshot saved: landing_page.png');
  
  // Check 1: Decorative gradient bar at the top
  try {
    const gradientElements = await page.$$('[class*="gradient"], [style*="gradient"]');
    if (gradientElements.length === 0) {
      issues.push('❌ No decorative gradient bar found at the top');
    } else {
      console.log('✓ Gradient bar found');
    }
  } catch (e) {
    issues.push(`❌ Error checking gradient bar: ${e.message}`);
  }
  
  // Check 2: "Goal-Setting Framework" label
  try {
    const frameworkLabel = await page.$('text/Goal-Setting Framework');
    if (frameworkLabel) {
      console.log('✓ "Goal-Setting Framework" label found');
    } else {
      issues.push('❌ "Goal-Setting Framework" label not found');
    }
  } catch (e) {
    issues.push(`❌ Error checking framework label: ${e.message}`);
  }
  
  // Check 3: "Harada Method" title (may be split across lines)
  try {
    const haradaText = await page.evaluate(() => {
      const h1 = document.querySelector('h1');
      return h1 ? h1.textContent.trim() : null;
    });
    
    if (haradaText && haradaText.includes('Harada') && haradaText.includes('Method')) {
      console.log('✓ "Harada Method" title found');
      
      // Check if "Method" is in a lighter color (span with opacity/muted class)
      const methodSpan = await page.$('h1 span');
      if (methodSpan) {
        const opacity = await page.evaluate(el => {
          const style = window.getComputedStyle(el);
          return style.opacity || style.color;
        }, methodSpan);
        console.log(`  "Method" styling: ${opacity}`);
        console.log('✓ "Method" appears to be in a lighter color');
      }
      
      // Check font size
      const h1 = await page.$('h1');
      const fontSize = await page.evaluate(el => {
        return window.getComputedStyle(el).fontSize;
      }, h1);
      console.log(`  Title font size: ${fontSize}`);
    } else {
      issues.push('❌ "Harada Method" title not found');
    }
  } catch (e) {
    issues.push(`❌ Error checking title: ${e.message}`);
  }
  
  // Check 4: "How it works" section with 3 steps
  try {
    const howItWorks = await page.$('text/How it works');
    if (howItWorks) {
      console.log('✓ "How it works" section found');
      
      // Check for steps
      const step01 = await page.$('text/01');
      const step02 = await page.$('text/02');
      const step03 = await page.$('text/03');
      
      const stepsFound = [];
      if (step01) stepsFound.push('01');
      if (step02) stepsFound.push('02');
      if (step03) stepsFound.push('03');
      
      if (stepsFound.length === 3) {
        console.log(`✓ All 3 steps found: ${stepsFound.join(', ')}`);
      } else {
        issues.push(`❌ Not all steps found. Found: ${stepsFound.join(', ') || 'none'}`);
      }
    } else {
      issues.push('❌ "How it works" section not found');
    }
  } catch (e) {
    issues.push(`❌ Error checking steps: ${e.message}`);
  }
  
  // Check 5: "Your Charts" section (only appears if charts exist)
  try {
    // Look for chart links first
    const chartLinks = await page.$$('a[href*="/chart/"]');
    
    if (chartLinks.length > 0) {
      console.log('✓ "Your Charts" section found');
      console.log(`✓ Found ${chartLinks.length} saved chart(s)`);
      
      if (issues.length > 0) {
        console.log('\n=== Landing Page Issues ===');
        issues.forEach(issue => console.log(issue));
      }
      
      return chartLinks;
    } else {
      console.log('⚠ "Your Charts" section not visible (no charts saved yet)');
      console.log('  Note: This section only appears when charts exist');
      
      if (issues.length > 0) {
        console.log('\n=== Landing Page Issues ===');
        issues.forEach(issue => console.log(issue));
      }
      
      return [];
    }
  } catch (e) {
    issues.push(`❌ Error checking charts section: ${e.message}`);
    
    if (issues.length > 0) {
      console.log('\n=== Landing Page Issues ===');
      issues.forEach(issue => console.log(issue));
    }
    
    return [];
  }
}

async function testChartView(page, chartUrl) {
  console.log('\n=== Testing Chart View ===');
  const issues = [];
  
  // Navigate to chart
  await page.goto(chartUrl, { waitUntil: 'networkidle0' });
  await page.waitForTimeout(1000); // Extra wait for rendering
  
  // Take screenshot
  await page.screenshot({ path: 'chart_view.png', fullPage: true });
  console.log('✓ Screenshot saved: chart_view.png');
  
  // Check 1: Grid with rounded border and shadow
  try {
    const grid = await page.$('[class*="grid"], table, [role="grid"]');
    if (grid) {
      console.log('✓ Grid found');
      
      const styles = await page.evaluate(el => {
        const computed = window.getComputedStyle(el);
        return {
          borderRadius: computed.borderRadius,
          boxShadow: computed.boxShadow
        };
      }, grid);
      
      if (styles.borderRadius && styles.borderRadius !== '0px') {
        console.log(`✓ Grid has rounded border: ${styles.borderRadius}`);
      } else {
        issues.push('❌ Grid does not have rounded border');
      }
      
      if (styles.boxShadow && styles.boxShadow !== 'none') {
        console.log(`✓ Grid has shadow: ${styles.boxShadow}`);
      } else {
        issues.push('❌ Grid does not have shadow');
      }
    } else {
      issues.push('❌ Grid not found');
    }
  } catch (e) {
    issues.push(`❌ Error checking grid: ${e.message}`);
  }
  
  // Check 2: Cells have proper coloring
  try {
    const cells = await page.$$('td, [role="gridcell"]');
    if (cells.length > 0) {
      console.log(`✓ Found ${cells.length} cells`);
      
      // Check first 10 cells for background color
      const cellsToCheck = cells.slice(0, Math.min(10, cells.length));
      let coloredCells = 0;
      
      for (const cell of cellsToCheck) {
        const bgColor = await page.evaluate(el => {
          return window.getComputedStyle(el).backgroundColor;
        }, cell);
        
        if (bgColor && bgColor !== 'rgba(0, 0, 0, 0)' && bgColor !== 'transparent') {
          coloredCells++;
        }
      }
      
      if (coloredCells > 0) {
        console.log(`✓ Cells have coloring (${coloredCells}/${cellsToCheck.length} checked cells have background color)`);
      } else {
        issues.push('❌ Cells do not appear to have proper coloring');
      }
    } else {
      issues.push('❌ No cells found in grid');
    }
  } catch (e) {
    issues.push(`❌ Error checking cell coloring: ${e.message}`);
  }
  
  // Check 3: Block borders
  try {
    const cells = await page.$$('td, [role="gridcell"]');
    if (cells.length >= 9) {
      // Check border styles on cells at block boundaries
      const cellBorders = [];
      for (const i of [0, 3, 6]) {
        if (i < cells.length) {
          const border = await page.evaluate(el => {
            return window.getComputedStyle(el).borderWidth;
          }, cells[i]);
          cellBorders.push(border);
        }
      }
      
      console.log(`✓ Cell border widths sampled: ${cellBorders.join(', ')}`);
      console.log('  (Manual verification needed for block vs inner border darkness)');
    } else {
      console.log('⚠ Not enough cells to verify block borders');
    }
  } catch (e) {
    issues.push(`❌ Error checking block borders: ${e.message}`);
  }
  
  if (issues.length > 0) {
    console.log('\n=== Chart View Issues ===');
    issues.forEach(issue => console.log(issue));
  }
  
  return issues;
}

async function main() {
  const browser = await puppeteer.launch({
    headless: true,
    args: ['--no-sandbox', '--disable-setuid-sandbox']
  });
  
  const page = await browser.newPage();
  let chartLinks = [];
  
  try {
    // Test landing page
    chartLinks = await testLandingPage(page);
    
    // Test chart view if charts exist
    if (chartLinks && chartLinks.length > 0) {
      const chartUrl = await page.evaluate(el => el.href, chartLinks[0]);
      console.log(`\nNavigating to chart: ${chartUrl}`);
      await testChartView(page, chartUrl);
    } else {
      console.log('\n⚠ No charts available to test chart view');
      console.log('Please create a chart first to test the chart view.');
    }
  } finally {
    await browser.close();
  }
  
  console.log('\n=== Testing Complete ===');
  console.log('Screenshots saved:');
  console.log('  - landing_page.png');
  if (chartLinks && chartLinks.length > 0) {
    console.log('  - chart_view.png');
  }
}

main().catch(err => {
  console.error('Error:', err);
  process.exit(1);
});

File: tests/test_visual_check.py
#!/usr/bin/env python3
"""
Visual testing script for Harada Method application.
Tests landing page and chart view for visual elements.
"""

from playwright.sync_api import sync_playwright
import sys

def test_landing_page(page):
    """Test landing page visual elements"""
    print("\n=== Testing Landing Page ===")
    issues = []
    
    # Navigate to landing page
    page.goto('http://localhost:3000')
    page.wait_for_load_state('networkidle')
    
    # Take screenshot
    page.screenshot(path='landing_page.png', full_page=True)
    print("✓ Screenshot saved: landing_page.png")
    
    # Check 1: Decorative gradient bar at the top
    gradient_bar = page.locator('div').filter(has_text='').first
    try:
        # Look for a gradient element at the top
        gradient_elements = page.locator('[class*="gradient"], [style*="gradient"]').all()
        if len(gradient_elements) == 0:
            issues.append("❌ No decorative gradient bar found at the top")
        else:
            print("✓ Gradient bar found")
    except Exception as e:
        issues.append(f"❌ Error checking gradient bar: {str(e)}")
    
    # Check 2: "Goal-Setting Framework" label
    try:
        framework_label = page.get_by_text("Goal-Setting Framework", exact=False)
        if framework_label.count() > 0:
            print("✓ 'Goal-Setting Framework' label found")
        else:
            issues.append("❌ 'Goal-Setting Framework' label not found")
    except Exception as e:
        issues.append(f"❌ Error checking framework label: {str(e)}")
    
    # Check 3: "Harada Method" title with "Method" in lighter color
    try:
        harada_title = page.get_by_text("Harada Method", exact=False)
        if harada_title.count() > 0:
            print("✓ 'Harada Method' title found")
            # Check if it's large (we can't easily verify color without inspecting styles)
            title_element = harada_title.first
            font_size = title_element.evaluate('el => window.getComputedStyle(el).fontSize')
            print(f"  Title font size: {font_size}")
        else:
            issues.append("❌ 'Harada Method' title not found")
    except Exception as e:
        issues.append(f"❌ Error checking title: {str(e)}")
    
    # Check 4: "How it works" section with 3 steps
    try:
        how_it_works = page.get_by_text("How it works", exact=False)
        if how_it_works.count() > 0:
            print("✓ 'How it works' section found")
            
            # Check for steps 01, 02, 03
            step_01 = page.get_by_text("01", exact=False)
            step_02 = page.get_by_text("02", exact=False)
            step_03 = page.get_by_text("03", exact=False)
            
            steps_found = []
            if step_01.count() > 0:
                steps_found.append("01")
            if step_02.count() > 0:
                steps_found.append("02")
            if step_03.count() > 0:
                steps_found.append("03")
            
            if len(steps_found) == 3:
                print(f"✓ All 3 steps found: {', '.join(steps_found)}")
            else:
                issues.append(f"❌ Not all steps found. Found: {', '.join(steps_found) if steps_found else 'none'}")
        else:
            issues.append("❌ 'How it works' section not found")
    except Exception as e:
        issues.append(f"❌ Error checking steps: {str(e)}")
    
    # Check 5: "Your Charts" section
    try:
        your_charts = page.get_by_text("Your Charts", exact=False)
        if your_charts.count() > 0:
            print("✓ 'Your Charts' section found")
            
            # Look for chart links/cards
            chart_links = page.locator('a[href*="/chart/"]').all()
            if len(chart_links) > 0:
                print(f"✓ Found {len(chart_links)} saved chart(s)")
                return chart_links
            else:
                print("⚠ No saved charts found (this may be expected)")
                return []
        else:
            issues.append("❌ 'Your Charts' section not found")
            return []
    except Exception as e:
        issues.append(f"❌ Error checking charts section: {str(e)}")
        return []
    
    if issues:
        print("\n=== Landing Page Issues ===")
        for issue in issues:
            print(issue)
    
    return []

def test_chart_view(page, chart_url):
    """Test chart view visual elements"""
    print("\n=== Testing Chart View ===")
    issues = []
    
    # Navigate to chart
    page.goto(chart_url)
    page.wait_for_load_state('networkidle')
    page.wait_for_timeout(1000)  # Extra wait for rendering
    
    # Take screenshot
    page.screenshot(path='chart_view.png', full_page=True)
    print("✓ Screenshot saved: chart_view.png")
    
    # Check 1: Grid with rounded border and shadow
    try:
        # Look for grid container
        grid = page.locator('[class*="grid"], table, [role="grid"]').first
        if grid.count() > 0:
            print("✓ Grid found")
            
            # Check border radius and shadow
            border_radius = grid.evaluate('el => window.getComputedStyle(el).borderRadius')
            box_shadow = grid.evaluate('el => window.getComputedStyle(el).boxShadow')
            
            if border_radius and border_radius != '0px':
                print(f"✓ Grid has rounded border: {border_radius}")
            else:
                issues.append("❌ Grid does not have rounded border")
            
            if box_shadow and box_shadow != 'none':
                print(f"✓ Grid has shadow: {box_shadow}")
            else:
                issues.append("❌ Grid does not have shadow")
        else:
            issues.append("❌ Grid not found")
    except Exception as e:
        issues.append(f"❌ Error checking grid: {str(e)}")
    
    # Check 2: Cells have proper coloring
    try:
        cells = page.locator('td, [role="gridcell"]').all()
        if len(cells) > 0:
            print(f"✓ Found {len(cells)} cells")
            
            # Check a few cells for background color
            colored_cells = 0
            for i, cell in enumerate(cells[:10]):  # Check first 10 cells
                bg_color = cell.evaluate('el => window.getComputedStyle(el).backgroundColor')
                if bg_color and bg_color not in ['rgba(0, 0, 0, 0)', 'transparent']:
                    colored_cells += 1
            
            if colored_cells > 0:
                print(f"✓ Cells have coloring ({colored_cells}/{min(10, len(cells))} checked cells have background color)")
            else:
                issues.append("❌ Cells do not appear to have proper coloring")
        else:
            issues.append("❌ No cells found in grid")
    except Exception as e:
        issues.append(f"❌ Error checking cell coloring: {str(e)}")
    
    # Check 3: Block borders (every 3 cells) darker than inner borders
    try:
        # This is complex to verify programmatically, so we'll do a basic check
        cells = page.locator('td, [role="gridcell"]').all()
        if len(cells) >= 9:  # Need at least 9 cells to check block borders
            # Check border styles on a few cells
            cell_borders = []
            for i in [0, 3, 6]:  # Check cells at block boundaries
                if i < len(cells):
                    border = cells[i].evaluate('el => window.getComputedStyle(el).borderWidth')
                    cell_borders.append(border)
            
            print(f"✓ Cell border widths sampled: {', '.join(cell_borders)}")
            print("  (Manual verification needed for block vs inner border darkness)")
        else:
            print("⚠ Not enough cells to verify block borders")
    except Exception as e:
        issues.append(f"❌ Error checking block borders: {str(e)}")
    
    if issues:
        print("\n=== Chart View Issues ===")
        for issue in issues:
            print(issue)
    
    return issues

def main():
    with sync_playwright() as p:
        browser = p.chromium.launch(headless=True)
        page = browser.new_page()
        
        try:
            # Test landing page
            chart_links = test_landing_page(page)
            
            # Test chart view if charts exist
            if chart_links:
                # Get the first chart URL
                chart_url = chart_links[0].get_attribute('href')
                if not chart_url.startswith('http'):
                    chart_url = f"http://localhost:3000{chart_url}"
                
                print(f"\nNavigating to chart: {chart_url}")
                test_chart_view(page, chart_url)
            else:
                print("\n⚠ No charts available to test chart view")
                print("Please create a chart first to test the chart view.")
        
        finally:
            browser.close()
        
        print("\n=== Testing Complete ===")
        print("Screenshots saved:")
        print("  - landing_page.png")
        if chart_links:
            print("  - chart_view.png")

if __name__ == '__main__':
    main()
